internal/heartbeat/heartbeat.go

```diff
@@ -195,12 +230,19 @@ func (s *Heartbeat) Init(config models.DeviceConfigurationMessage) error {
 }
 
 func (s *Heartbeat) Update(config models.DeviceConfigurationMessage) error {
+	s.tickerLock.Lock()
 	periodSeconds := s.getInterval(*config.Configuration)
-	log.Infof("reconfiguring ticker with interval: %v. DeviceID: %s", periodSeconds, s.data.workloadManager.GetDeviceID())
-	if s.ticker != nil {
-		s.ticker.Stop()
+	if s.previousPeriodSeconds <= 0 || s.previousPeriodSeconds != periodSeconds {
+		s.log.Debugf("Heartbeat configuration update: periodSeconds changed from %d to %d; Device ID: %s", s.previousPeriodSeconds, periodSeconds, s.data.workloadManager.GetDeviceID())
+		s.tickerLock.Unlock()
```

Comment:
maybe do defer on the line 233 is the best option tbh.

Reply:
Had to unlock here to avoid deadlock when call s.initTicker

Reply:
:+1: